File: SimSim/Actions.m
//
//  Actions.m
//  SimSim
//
//  Created by Daniil Smelov on 05.04.18.
//  Copyright © 2018 Daniil Smelov. All rights reserved.
//

#import "Actions.h"
#import "Tools.h"
#import "Settings.h"
#import "FileManagerSupport/CommanderOne.h"

@implementation Actions

//----------------------------------------------------------------------------
+ (void) copyToPasteboard:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    NSPasteboard* pasteboard = [NSPasteboard generalPasteboard];
    
    [pasteboard declareTypes:@[ NSPasteboardTypeString ] owner:nil];
    [pasteboard setString:path forType:NSPasteboardTypeString];
}

//----------------------------------------------------------------------------
+ (void) takeScreenshot:(id)sender
{
    NSArray* windows = (NSArray *)CFBridgingRelease(CGWindowListCopyWindowInfo(kCGWindowListExcludeDesktopElements, kCGNullWindowID));
    
    for(NSDictionary *window in windows)
    {
        NSString* windowOwner = window[(NSString*)kCGWindowOwnerName];
        NSString* windowName = window[(NSString*)kCGWindowName];
        
        if ([windowOwner containsString:@"Simulator"] &&
            ([windowName containsString:@"iOS"] || [windowName containsString:@"watchOS"] || [windowName containsString:@"tvOS"]))
        {
            NSNumber* windowID = window[(NSString*)kCGWindowNumber];
            
            NSString *dateComponents = @"yyyyMMdd_HHmmss_SSSS";
            NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
            [dateFormatter setTimeZone:[NSTimeZone localTimeZone]];
            [dateFormatter setDateFormat:dateComponents];
            
            NSDate *date = [NSDate date];
            NSString *dateString = [dateFormatter stringFromDate:date];
            
            NSString* screenshotPath =
            [NSString stringWithFormat:@"%@/Desktop/Screen Shot at %@.png", [Tools homeDirectoryPath], dateString];
            
            CGRect bounds;
            CGRectMakeWithDictionaryRepresentation((CFDictionaryRef)window[(NSString*)kCGWindowBounds], &bounds);
            
            CGImageRef image = CGWindowListCreateImage(bounds, kCGWindowListOptionIncludingWindow, (CGWindowID)[windowID intValue], kCGWindowImageDefault);
            NSBitmapImageRep *bitmap = [[NSBitmapImageRep alloc] initWithCGImage:image];
            
            NSData *data = [bitmap representationUsingType: NSPNGFileType properties:@{}];
            [data writeToFile: screenshotPath atomically:NO];
            
            CGImageRelease(image);
        }
    }
}

//----------------------------------------------------------------------------
+ (void) resetFolder:(NSString*)folder inRoot:(NSString*)root
{
    NSString* path = [root stringByAppendingPathComponent:folder];
    
    NSFileManager* fm = [NSFileManager new];
    NSDirectoryEnumerator* en = [fm enumeratorAtPath:path];
    NSError* error = nil;
    BOOL result = NO;
    
    NSString* file;
    
    while (file = [en nextObject])
    {
        result = [fm removeItemAtPath:[path stringByAppendingPathComponent:file] error:&error];
        if (!result && error)
        {
            NSLog(@"Something went wrong: %@", error);
        }
    }
}

//----------------------------------------------------------------------------
+ (void) resetApplication:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    [self resetFolder:@"Documents" inRoot:path];
    [self resetFolder:@"Library" inRoot:path];
    [self resetFolder:@"tmp" inRoot:path];
}

//----------------------------------------------------------------------------
+ (void) openInFinder:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    [[NSWorkspace sharedWorkspace] openFile:path withApplication:@"Finder"];
}

//----------------------------------------------------------------------------
+ (void) openInTerminal:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    [[NSWorkspace sharedWorkspace] openFile:path withApplication:@"Terminal"];
}

//----------------------------------------------------------------------------
+ (void) openIniTerm:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    [[NSWorkspace sharedWorkspace] openFile:path withApplication:@"iTerm"];
}

//----------------------------------------------------------------------------
+ (void) openInCommanderOne:(id)sender
{
    NSString* path = (NSString*)[sender representedObject];
    
    [CommanderOne openInCommanderOne:path];
}

//----------------------------------------------------------------------------
+ (void) exitApp:(id)sender
{
    [[NSApplication sharedApplication] terminate:self];
}

//----------------------------------------------------------------------------
+ (void) handleStartAtLogin:(id)sender
{
    BOOL isEnabled = [[sender representedObject] boolValue];
    
    [Settings setStartAtLoginEnabled:!isEnabled];
    
    [sender setRepresentedObject:@(!isEnabled)];
    
    if (isEnabled)
    {
        [sender setState:NSOffState];
    }
    else
    {
        [sender setState:NSOnState];
    }
}

//----------------------------------------------------------------------------
+ (void) aboutApp:(id)sender
{
    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:@"https://github.com/dsmelov/simsim"]];
}


@end

File: SimSim/Actions.swift
//
//  Actions.swift
//  SimSim
//
//  Created by Daniil Smelov on 13/04/2018.
//  Copyright © 2018 Daniil Smelov. All rights reserved.
//

import Foundation
import Cocoa

//----------------------------------------------------------------------------
class Actions: NSObject
{
    //----------------------------------------------------------------------------
    @objc
    class func copy(toPasteboard sender: NSMenuItem)
    {
        let path = sender.representedObject as! String
        let pasteboard = NSPasteboard.general
        pasteboard.declareTypes([NSPasteboard.PasteboardType.string], owner: nil)
        pasteboard.setString(path, forType: NSPasteboard.PasteboardType.string)
    }

    //----------------------------------------------------------------------------
    class func resetFolder(_ folder: String, inRoot root: String!)
    {
        let pathUrl = URL(fileURLWithPath: root).appendingPathComponent(folder)
        let fm = FileManager()
        do
        {
            try fm.removeItem(at: pathUrl)
        }
        catch let error as NSError
        {
            print("Ooops! Something went wrong: \(error)")
        }
    }

    //----------------------------------------------------------------------------
    @objc
    class func resetApplication(_ sender: NSMenuItem)
    {
        let folders = ["Documents", "Library", "tmp"]
        
        for victim in folders
        {
            self.resetFolder(victim, inRoot: sender.representedObject as? String)
        }
    }

    //----------------------------------------------------------------------------
    class func open(item: NSMenuItem, with: String, appendingPath: String = "")
    {
        guard var path = item.representedObject as? String else { return }
        path += appendingPath
        NSWorkspace.shared.openFile(path, withApplication: with)
    }
    
    //----------------------------------------------------------------------------
    @objc
    class func open(inFinder sender: NSMenuItem)
    {
        open(item: sender, with: Constants.Actions.finder)
    }
    
    //----------------------------------------------------------------------------
    @objc
    class func open(inTerminal sender: NSMenuItem)
    {
        open(item: sender, with: Constants.Actions.terminal)
    }

    //----------------------------------------------------------------------------
    @objc
    class func openIniTerm(_ sender: NSMenuItem)
    {
        open(item: sender, with: Constants.Actions.iTerm)
    }
    
    //----------------------------------------------------------------------------
    @objc
    class func open(inCommanderOne sender: NSMenuItem)
    {
        // For some reason Commander One opens not the last folder in path
        open(item: sender, with: Constants.Actions.commanderOne, appendingPath: "Library/")
    }
    
    //----------------------------------------------------------------------------
    @objc
    class func exitApp(_ sender: NSMenuItem)
    {
        NSApplication.shared.terminate(self)
    }

    //----------------------------------------------------------------------------
    @objc
    class func handleStart(atLogin sender: NSMenuItem)
    {
        let isEnabled: Bool = sender.representedObject as! Bool
        
        Settings.isStartAtLoginEnabled = !isEnabled
        sender.representedObject = !isEnabled
        
        if isEnabled
        {
            sender.state = .off
        }
        else
        {
            sender.state = .on
        }
    }

    //----------------------------------------------------------------------------
    @objc
    class func aboutApp(_ sender: NSMenuItem)
    {
        NSWorkspace.shared.open(URL(string: Constants.githubUrl)!)
    }

    //----------------------------------------------------------------------------
    @objc
    class func openIn(withModifier sender: NSMenuItem)
    {
        guard let event = NSApp.currentEvent else { return }
        
        if event.modifierFlags.contains(.option)
        {
            Actions.open(inTerminal: sender)
        }
        else if event.modifierFlags.contains(.control)
        {
            if Tools.commanderOneAvailable()
            {
                Actions.open(inCommanderOne: sender)
            }
        }
        else
        {
            Actions.open(inFinder: sender)
        }
    }

}


























